### PR TITLE
added tests for math operators

### PR DESCRIPTION
**Pull request by annzhukovets**

### CODE REVIEW

File: tests/apps/test_operators.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+        cxt = tc.Context()
+        w_tc = self.w1_tc.exp()
+        y_tc = self.x_tc*w_tc
+        cxt.result = form_of(y_tc).gradients(tc.tensor.Dense.ones(y_tc.shape))[hex_id(self.w1_tc)]
+        w1_tc_grad = load_np(HOST.post(ENDPOINT, cxt))
+
+        self.assertTrue((abs(w1_tc_grad-[t.numpy() for t in w1_torch_grad]) < 0.0001).all())
+
+    # TODO: make it works (chain rule)
```
Comment (haydnv) [inline tests/apps/test_operators.py:129]: There were a few times when I was working in the neural_net.py repo that I decided I must be missing the chain rule, and each time it turned out that the problem was something else (e.g. getting a partial derivative wrong) and the chain rule was being computed implicitly by calling `Operator.derivative` recursively. I think you should step through the result of each `Operator` in the graph (e.g. `cxt.result = w_tc`, then `cxt.result = y_tc`, etc) to identify the point of divergence. It would be clearer to construct the `Operator` graph in a separate `Op`, e.g. `@tc.post my_calculation(...):` and check the return values one-by-one. If it seems like you do need to apply the chain rule explicitly, you can try defining it just in the `Exp` and `MatMul` operators, to check if the result matches your expectations.

**Review by haydnv — CHANGES_REQUESTED** [review]

File: tests/apps/test_operators.py (hunk 1 above)
Comment (haydnv) [reply tests/apps/test_operators.py:129]: Here's a simple example of calculating the chain rule:
```python
class Exp(tc.math.operator.Unary):
    @staticmethod
    def diff(x, wrt=None):
        return Exp(x)

    def derivative(self, variable):
        if is_operator(self.subject):
            return self.diff(self.subject, variable) * derivative_of(self.subject, variable)

        return self.diff(self.subject, variable)

def is_operator(state):
    return isinstance(form_of(state), tc.math.Operator)
```

You can follow this same pattern to implement `MatMul` and `Add` in the test, e.g. `y_tc = Add(MatMul(self.x_tc, self.w1_tc), self.w1_tc)`, to check if your intuition about the chain rule is correct.

Incidentally, did you mean to use `b1_torch` and `b1_tc` instead of `w1_torch` and `w1_tc`?

File: tests/apps/test_operators.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+# LIB_URI = tc.uri(tc.math.linalg.LinearAlgebra)
+TENSOR_URI = str(tc.uri(tc.tensor.Dense))
+HOST = tc.host.Host('http://127.0.0.1:8702')
+ENDPOINT = '/transact/hypothetical'
+
+
+class OperatorTests(unittest.TestCase):
+    # @classmethod
+    # def setUpClass(cls):
+    #     cls.host = testutils.start_host("test_neural_net", tc.math.linalg.LinearAlgebra(), wait_time=2)
```
Comment (haydnv) [inline tests/apps/test_operators.py:17]: If you rebase or `merge main` you can leave out the second argument here

File: tests/apps/test_operators.py (hunk 1 above)
Comment (haydnv) [reply tests/apps/test_operators.py:297]: You can follow up on this in a separate PR so we can merge in the trig operator gradients now

File: tests/apps/testutils.py [hunk 3]
```diff
@@ -46,7 +47,7 @@ def start(self, wait_time):
         print(f"running new Docker container with image {image.id}")
         self.container = self.client.containers.run(
             image.id, cmd,
-            network_mode=DOCKER_NETWORK_MODE, volumes=[f"{self.config_dir}:/{CONFIG}"], detach=True)
+            network_mode=DOCKER_NETWORK_MODE, volumes=[f"{self.config_dir}:/{CONFIG}"], detach=True, ports=DOCKER_PORTS)
```
Comment (haydnv) [inline tests/apps/testutils.py:50]: note that some tests start hosts on other ports also--you can use `flags.get("http_port", DEFAULT_PORT)` to see which port to map

**Review by haydnv — APPROVED** [review]

File: client/tinychain/math/operator.py [hunk 4]
```diff
@@ -280,6 +311,16 @@ def backward(self, variable):
         subject = derivative_of(self.subject, variable)
         return subject - self.subject.sin()
 
+    def gradients(self, loss):
+        from ..ml.optimizer import Variable
+
+        grad = -self.subject.sin() * loss
+
+        if isinstance(self.subject, Variable):
```
Comment (haydnv) [inline client/tinychain/math/operator.py:319]: Could this logic go in a parent class? e.g.
```python
class Trig(Unary):
  # ...
  def gradients(self, loss):
        from ..ml.optimizer import Variable
        if isinstance(self.subject, Variable):
          return self.subject.invert(loss)
        elif isinstance(form_of(self.subject), Operator):
          return form_of(self.subject).gradients(loss)
        else:
          return GradientDict({})

class Cos(Trig):
  # ...
  def gradients(self, loss):
    grad = -self.subject.sin() * loss
    return Trig.gradients(grad)
```

File: tests/apps/test_operators.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+import torch
+from torch.autograd import grad
+from tinychain.util import form_of, hex_id
+
+TENSOR_URI = str(tc.uri(tc.tensor.Dense))
+HOST = tc.host.Host('http://127.0.0.1:8702')
+ENDPOINT = '/transact/hypothetical'
+
+
+class OperatorTests(unittest.TestCase):
```
Comment (haydnv) [inline tests/apps/test_operators.py:14]: I think this should actually go in `tests/client` since it's a test of functionality implemented in the Python client and doesn't need to load a cluster configuration

**Review by haydnv — CHANGES_REQUESTED** [review]
Great work getting `testMultipleFunctions` to pass! But tt looks like these changes break `testCNN` and `testDNN` in `NeuralNetTests`. For `testDNN` I think you can just add logic somewhere to handle the batch dimension (this might just be an operator called `Reduce` that gets returned by `Tensor.sum(0)`) but `testCNN` may require more custom `Operator` definitions.

**Comment by haydnv** [comment]
After looking at this for longer, I think we need to enforce a rule that the output of `Operator.backward` always has the same shape as the output of `Operator.forward` and gradients calculated in `Operator.gradients` always have the same shape as the `subject` of the `Operator` (and the `args` of a `Dual` operator, where applicable). This will require a function like `unbroadcast` that sums over the broadcasted axes of the `loss` passed to `Add`, `Div`, `Mul`, and `Sub` with respect to the operator's `subject` and `args`.

Also, I added a `Slice` operator, PTAL: 7114aebe

**Comment by haydnv** [comment]
After looking at this for longer, I think we need to enforce a rule that the output of `Operator.backward` always has the same shape as the output of `Operator.forward` and gradients calculated in `Operator.gradients` always have the same shape as the `subject` of the `Operator` (and the `args` of a `Dual` operator, where applicable). This will require a function like `unbroadcast` that sums over the broadcasted axes of the `loss` passed to `Add`, `Div`, `Mul`, and `Sub` with respect to the operator's `subject` and `args`.

Also, I added a `Slice` operator, PTAL: 7114aebe

**Comment by haydnv** [comment]
Closing this since I believe all these changes were merged in as part of #140 and this is an older branch, but please reopen if I missed anything (or, preferably, send a new PR with the changes on a newer branch).

**Comment by haydnv** [comment]
Sorry, I just looked at this again and saw the `Sum` operator! I sent you #141 on a new branch to review and test.
